Improve warning message about unreachable feed items.

mbwgh opened the pull request:
Resolves #1548. Better late than never.

The unreachable items' feed urls are printed via the `vectorofstring_callback`, which makes them visible in the logfile if the users wants to see them, albeit with a rather generic log message.

The warning message is as suggested in the issue.

I am unsure whether the translation step has been done correctly, the steps necessary weren't really obvious so I ran `make msgmerge` and adjusted the German translation at least.

Minoru commented:
Sorry, I've got COVID and can't give this the proper attention right now. Also, the next release is in two weeks, which we usually spend in a "feature freeze" of sorts -- no changes that touch the i18n files are allowed while the translators are updating them. So this will probably get reviewed in about a week from now, and will be released in December. Sorry about that!

mbwgh commented:
> Sorry, I've got COVID and can't give this the proper attention right now. Also, the next release is in two weeks, which we usually spend in a "feature freeze" of sorts -- no changes that touch the i18n files are allowed while the translators are updating them. So this will probably get reviewed in about a week from now, and will be released in December. Sorry about that!

Ouch, get well soon!
No pressure from my end :)

Minoru commented on doc/configcommands.dsv:
```diff
@@ -10,7 +10,7 @@ bookmark-cmd||<command>||""||If set, then <command> will be used as bookmarking
 browser||<command>||%BROWSER, otherwise lynx||Set the browser command to use when opening an article in the browser. If the <<BROWSER,`BROWSER`>> environment variable is set, it will be used as the default browser, otherwise lynx will be used. For more information, see <<_using_browser,Using Browser>>.||browser "w3m %u"
 cache-file||<path>||"~/.newsboat/cache.db" or "~/.local/share/cache.db" (see "Files" section)||This configuration option sets the cache file. This is especially useful if the filesystem of your home directory doesn't support proper locking (e.g. NFS).||cache-file "/tmp/testcache.db"
-cleanup-on-quit||[yes/no]||yes||If set to `yes`, then the cache gets locked and superfluous feeds and items are removed, such as feeds that can't be found in the urls configuration file anymore.||cleanup-on-quit no
+cleanup-on-quit||[yes/no]||yes||If set to `yes`, then the cache gets locked and superfluous feeds and items are removed, such as feeds that can't be found in the urls configuration file anymore. Run `newsboat --cleanup` to do this manually. If you encounter a warning about unreachable feeds having been found, you may see the feed urls listed by creating a log file.||cleanup-on-quit no
```
> by creating a log file

There are two log files in Newsboat: the one created with `--log-file` on the command line, and the one created with the `error-log` setting. The former is more of a debug facility, while the latter is user-facing; I'd prefer to keep it that way. So how about writing this info into the error log instead? For that, you have to log at `Level::USERERROR`, and adjust the text above to mention `error-log`.

mbwgh replied:
Using `USERERROR`-level and mentioning the `error-log` option now.

Minoru commented on src/cache.cpp:
```diff
@@ -812,7 +812,7 @@ std::uint64_t Cache::cleanup_cache(std::vector<std::shared_ptr<RssFeed>> feeds,
 
 		std::string query = (
-				"SELECT count(rss) "
+				"SELECT distinct rss "
```
```suggestion
				"SELECT DISTINCT rss "
```

We put SQL keywords into uppercase to make the query easier to read.

mbwgh replied:
Fixed.

Minoru commented on src/controller.cpp:
```diff
@@ -561,8 +561,8 @@ int Controller::run(const CliArgsParser& args)
 			if (amt > 0u) {
-				std::cout << _("Unreachable feeds found, consider setting "
-						"`cleanup-on-quit yes` or run `newsboat --cleanup`")
+				std::cout << strprintf::fmt("%d unreachable feeds found. See "
+						"`cleanup-on-quit` in newsboat(1) for details.", amt)
```
```suggestion
				std::cout << strprintf::fmt(_("%d unreachable feeds found. See "
						"`cleanup-on-quit` in newsboat(1) for details."), amt)
```

Translatable strings have to be annotated with `_()`.

mbwgh replied:
Fixed.

Minoru reviewed (changes requested):
This is a good start, but I think it could be made a bit more user-friendly. Please see my comments below.

I suggest you drop the last two commits. We update .po files once every three months, before the release, so you shouldn't need to touch them at all. This will require a force-push, which is usually frowned upon as it makes me re-review everything, but this PR is small enough that it's okay. You also need to rebase onto current master to pick up fixes for FreeBSD CI. It looks like you know your way around Git, so shouldn't be a problem, but if you run into any trouble doing these manipulations, please comment and I'll help get you untangled.

mbwgh commented:
Rebased onto current master and dropped the last two commits as suggested.

Uses `LEVEL::USERERROR` now explicitly, albeit the callback handler used during query execution still does `DEBUG`-level logging and I left that as-is.

mbwgh commented:
Marking ready for review, since I am not sure if I am to blame for the tree failing CI jobs (one of them even fails with `internal compiler error`).

Minoru commented on src/controller.cpp:
```diff
@@ -555,13 +555,17 @@ int Controller::run(const CliArgsParser& args)
+					LOG(Level::USERERROR, "Unreachable feed found: %s", feed);
+				}
+
+				std::cout << strprintf::fmt(_("%d unreachable feeds found. See "
```
`cleanup_cache` returns a vector of feeds *and items*, so this message is misleading. Also, I think mixing URLs of feeds with URLs of items can be confusing. How about returning two separate vectors and displaying two separate counts?

mbwgh replied:
I am not sure if this is the case here.
An excerpt from my `cache.db`:

```sqlite
sqlite> select title, url, feedurl from rss_item limit 90000, 4;
title                                                         url                                                           feedurl
------------------------------------------------------------  ------------------------------------------------------------  ---------------------------
Ketchum, Idaho, Has Plenty of Available Jobs, but Workers Ca  https://www.wsj.com/articles/ketchum-idaho-has-plenty-of-ava  https://hnrss.org/frontpage
n’t Afford Housing                                            ilable-jobs-but-workers-cant-afford-housing-11625659200

Neurons Unexpectedly Encode Information in the Timing of The  https://www.quantamagazine.org/a-new-kind-of-information-cod  https://hnrss.org/frontpage
ir Firing                                                     ing-seen-in-the-human-brain-20210707/

Designing Our Serverless Engine: From Kubernetes to Nomad, F  https://www.koyeb.com/blog/the-koyeb-serverless-engine-from-  https://hnrss.org/frontpage
irecracker, and Kuma                                          kubernetes-to-nomad-firecracker-and-kuma

Game Gulf (2013)                                              https://the-magazine.org/11/game-gulf                         https://hnrss.org/frontpage
```

Since we only fetch `feedurl`s from `rss_item`, no actual items make it into this vector. In any case they would not satisfy the `NOT IN (feeds...)` criterion.

I think this is a good thing too, since the number of items would become pretty large if printed out in the user log, and printing two separate counters would go against the original goal of making the warning message less annoying.

mbwgh replied:
This means of course that the docstring for `cleanup_cache` was not correct :)

Minoru replied:
Ah, the docstring threw me off. Thanks for fixing it!

Minoru reviewed (changes requested):
Internal compiler error was probably an OOM kill. This happens sometimes because of the "greedy" containers we use.

The other two failures are real, please see my comment below.

Minoru commented on src/controller.cpp:
```diff
@@ -555,13 +555,17 @@ int Controller::run(const CliArgsParser& args)
+					LOG(Level::USERERROR, "Unreachable feed found: %s", feed);
+				}
+
+				std::cout << strprintf::fmt(_("%d unreachable feeds found. See "
+							"`cleanup-on-quit` in newsboat(1) for details."), unreachable_feeds.size())
```
(Hmm, I left a comment here, but it disappear. Reposting.)

This fails to compile on macOS because `strprintf::fmt` doesn't have an overload for the type that `std::vector::size` returns. Please cast it to `std::uint64_t` (should be enough :) and format it with `PRIu64`.

mbwgh replied:
I did so.

Minoru commented on po/newsboat.pot:
```diff
@@ -476,8 +476,7 @@ msgstr ""
 msgid ""
-"Unreachable feeds found, consider setting `cleanup-on-quit yes` or run "
-"`newsboat --cleanup`"
+"%d unreachable feeds found. See `cleanup-on-quit` in newsboat(1) for details."
```
Please don't touch this file at all, it will get updated automatically before the release.

mbwgh replied:
I reverted this change.

Minoru commented:
@mbwgh, sorry to pester you, but do you plan to finish this PR? It looks like it's almost ready, it'd be nice to finally merge it!

mbwgh commented:
> @mbwgh, sorry to pester you, but do you plan to finish this PR?

@Minoru Apologies, I absolutely do, just RL getting in the way. I'll take another stab at it this weekend hopefully :)

mbwgh commented:
Applied review suggestions, sans the thing about the feeds vs. items, about which I am hopefully not mistaken :)

Since you make a point of having good commit messages, I am not sure if you want a "clean" commit history, so please tell me if I should squash or rebase or if there's some other convention.

Minoru reviewed (approved):
Thanks! I was wrong about feeds vs. items, which means I blocked this pretty much for nothing; sorry about that!

I don't have requirements for clean history, but I sometimes squash it if it's *too* dirty to my taste :) This PR seems fine though, so I've enabled auto-merge now.

Thanks once again for the work!
